src/nvim/diff.c

```diff
@@ -2062,28 +2092,216 @@ static void run_linematch_algorithm(diff_T *dp)
+  } else if (diff_allignment == CHARMATCH || diff_allignment == WORDMATCH) {
+    dp->charmatchp = xmalloc(total_chars_length * sizeof(int));  // will hold results
+    dp->n_charmatch = total_chars_length;
+
+    bool lim_exceeded = false;
+    if (diff_allignment == CHARMATCH && total_chars_length > (size_t)chardiff_chars) {
+      lim_exceeded = true;
+    } else if (diff_allignment == WORDMATCH && total_word_count > (size_t)worddiff_words) {
+      lim_exceeded = true;
+    }
+
+    if (lim_exceeded == true) {
+      // do not run charmatch on the entire diff block
+      // we will attempt to run charmatch on the individual lines later
+      // for now, just initialize the result memory
+      for (size_t i = 0; i < total_chars_length; i++) {
+        dp->charmatchp[i] = -1;  // -1 indicates that algorithm has not yet ran
+      }
+    } else {
+      for (size_t i = 0; i < total_chars_length; i++) {
+        dp->charmatchp[i] = 0;  // default to not highlighted
+      }
+
+      // check is this a line that does not exist in other buffers?
+      // if so, highlight it as a 'newline', and we don't need to run the algorithm
+      bool newline = true;
+      for (size_t i = 0, c = 0; i < ndiffs; i++) {
+        if (diff_length[i] > 0) {
+          c++;
+        }
+        if (c > 1) {
+          newline = false;
+          break;
+        }
+      }
+
+      if (newline == true) {
+        for (size_t i = 0; i < total_chars_length; i++) {
+          dp->charmatchp[i] = 2;
+        }
+      } else {
+        int *decisions = NULL;
+        size_t decisions_length = linematch_nbuffers((const char **)diffbufs, diff_length, ndiffs,
+                                                     &decisions, 1, word_offset, word_offset_size);
+        for (size_t i = 0; i < decisions_length; i++) {
+          if (decisions[i] == (pow(2, (double)ndiffs) - 1)) {
+            // it's a comparison of all the buffers (don't highlight)
+            for (size_t j = 0; j < ndiffs; j++) {
+              for (size_t k = 0;
+                   k <
+                   (diff_allignment ==
+                    WORDMATCH ? word_offset_size[j][word_offset_result_index[j]] : 1); k++) {
+                size_t l = result_diff_start_pos[j]++;
+                dp->charmatchp[iwhite_index_offset ? iwhite_index_offset[l] + l : l] = 0;
+              }
+              word_offset_result_index[j]++;
+            }
+          } else {
+            // it's a skip in a single buffer (highlight as changed)
+            for (size_t j = 0; j < ndiffs; j++) {
+              if (decisions[i] & (1 << j)) {
+                for (size_t k = 0;
+                     k <
+                     (diff_allignment ==
+                      WORDMATCH ? word_offset_size[j][word_offset_result_index[j]] : 1); k++) {
+                  size_t l = result_diff_start_pos[j]++;
+                  dp->charmatchp[iwhite_index_offset ? iwhite_index_offset[l] + l : l] = 1;
+                }
```

Comment:
this for-loop is nearly identical to the one above. is there any opportunity for deduplication of the code in this PR, without over-abstracting things? that will help in the future.